OpenRA.Mods.Common/UpdateRules/Rules/20210321/UnhardcodeSquadManager.cs

```diff
@@ -11,33 +11,128 @@
+		public override IEnumerable<string> BeforeUpdateActors(ModData modData, List<MiniYamlNode> resolvedActors)
 		{
-			var addNodes = new List<MiniYamlNode>();
+			var aircraft = new List<string>();
+			var vips = new List<string>();
+
+			foreach (var actor in resolvedActors)
+			{
+				if (actor.Key.StartsWith('^'))
+					continue;
+
+				var isVip = false;
+				var isBuildable = false;
+				var isBuilding = false;
+				var isAircraft = false;
+				var isExcluded = false;
+				var canAttack = false;
+				var isKillable = false;
+
+				foreach (var trait in actor.Value.Nodes)
+				{
+					if (trait.IsRemoval())
+						continue;
+
+					if (trait.KeyMatches("Buildable"))
+					{
+						isBuildable = true;
+						continue;
+					}
+
+					if (trait.KeyMatches("Aircraft"))
+					{
+						isAircraft = true;
+						continue;
+					}
+
+					if (trait.KeyMatches("Health"))
+					{
+						isKillable = true;
+						continue;
+					}
+
+					if (vipsNames.Any(v => trait.KeyMatches(v)))
+					{
+						isVip = true;
+						continue;
+					}
+
+					if (buildings.Any(b => trait.KeyMatches(b)))
+					{
+						isBuilding = true;
+						continue;
+					}
+
+					if (excludedBuildings.Any(eb => trait.KeyMatches(eb)))
+					{
+						isExcluded = true;
+						continue;
+					}
+
+					if (attackBase.Any(ab => trait.KeyMatches(ab)))
+						canAttack = true;
+				}
 
-			var aircraft = modData.DefaultRules.Actors.Values.Where(a => a.HasTraitInfo<AircraftInfo>() && a.HasTraitInfo<AttackBaseInfo>()).Select(a => a.Name);
-			var airUnits = new MiniYamlNode("AirUnitsTypes", FieldSaver.FormatValue(aircraft.ToList()));
-			addNodes.Add(airUnits);
+				if (isAircraft && isBuildable && canAttack && isKillable)
+				{
+					var name = actor.Key.ToLower();
+					if (!aircraft.Contains(name))
+						aircraft.Add(name);
+				}
 
-			var vips = modData.DefaultRules.Actors.Values.Where(a => a.HasTraitInfo<HarvesterInfo>() || a.HasTraitInfo<BaseBuildingInfo>() || (a.HasTraitInfo<BuildingInfo>() && a.HasTraitInfo<BuildableInfo>() && !a.HasTraitInfo<LineBuildInfo>() && !a.HasTraitInfo<PlugInfo>())).Select(a => a.Name);
-			var protection = new MiniYamlNode("ProtectionTypes", FieldSaver.FormatValue(vips.ToList()));
-			addNodes.Add(protection);
+				if (isBuildable && isKillable && (isVip || (isBuilding && !isExcluded)))
+				{
+					var name = actor.Key.ToLower();
+					if (!vips.Contains(name))
+						vips.Add(name);
+				}
+			}
 
+			addNodes.Add(new MiniYamlNode("AirUnitsTypes", FieldSaver.FormatValue(aircraft)));
+			addNodes.Add(new MiniYamlNode("ProtectionTypes", FieldSaver.FormatValue(vips)));
+
+			yield break;
+		}
+
+		bool anyAdded = false;
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
 			foreach (var squadManager in actorNode.ChildrenMatching("SquadManagerBotModule"))
+			{
 				foreach (var addNode in addNodes)
 					squadManager.AddNode(addNode);
 
+				anyAdded = true;
```

Comment:
Same here. Something like
```suggestion
				foreach (var addNode in addNodes)
				{
					if (<squadManager has node with same key as addNode>)
						continue;

					squadManager.AddNode(addNode);
					anyAdded = true;
				}
```